governance/pyth_staking_sdk/src/utils/vesting.ts

```diff
@@ -0,0 +1,58 @@
+import type { UnlockSchedule, VestingSchedule } from "../types";
+
+export const getUnlockSchedule = (options: {
+  pythTokenListTime: bigint;
+  vestingSchedule: VestingSchedule;
+}): UnlockSchedule => {
+  const { vestingSchedule, pythTokenListTime } = options;
+
+  if (vestingSchedule.fullyVested) {
+    return [];
+  } else if (vestingSchedule.periodicVestingAfterListing) {
+    return getPeriodicUnlockSchedule({
+      balance: vestingSchedule.periodicVestingAfterListing.initialBalance,
+      numPeriods: vestingSchedule.periodicVestingAfterListing.numPeriods,
+      periodDuration:
+        vestingSchedule.periodicVestingAfterListing.periodDuration,
+      startDate: pythTokenListTime,
+    });
+  } else if (vestingSchedule.periodicVesting) {
+    return getPeriodicUnlockSchedule({
+      balance: vestingSchedule.periodicVesting.initialBalance,
+      numPeriods: vestingSchedule.periodicVesting.numPeriods,
+      periodDuration: vestingSchedule.periodicVesting.periodDuration,
+      startDate: vestingSchedule.periodicVesting.startDate,
+    });
+  }
+  assertNever(vestingSchedule);
+};
+
+export const getPeriodicUnlockSchedule = (options: {
+  balance: bigint;
+  startDate: bigint;
+  periodDuration: bigint;
+  numPeriods: bigint;
+}): UnlockSchedule => {
+  const { balance, startDate, periodDuration, numPeriods } = options;
+
+  const unlockSchedule: UnlockSchedule = [];
+  const currentTimeStamp = new Date().getTime() / 1000;
+
+  for (let i = 0; i < numPeriods; i++) {
+    const unlockTimeStamp =
+      Number(startDate) + Number(periodDuration) * (i + 1);
+    if (currentTimeStamp < unlockTimeStamp) {
+      unlockSchedule.push({
+        date: new Date(unlockTimeStamp * 1000),
+        amount: balance / numPeriods,
+      });
+    }
+  }
+
+  return unlockSchedule;
+};
+
+// Utility function to assert a value is never
+function assertNever(value: never): never {
+  throw new Error(`Unexpected value: ${value}`);
+}
```

Comment:
There's no real reason to wrap this in a function, IMO it just adds indirection to what's actually happening and it's easier to just read a throw at the call site.

What I do often do, however, is to create a custom error type to make it really easy to know what error I'm catching without doing string matching:

```ts
class UnexpectedValueError extends Error {
    constructor(value: unknown) {
        super(`Unexpected value: ${value}`);
    }
}
```

Reply:
n/a anymore